lineapy/graph_reader/graph_refactorer.py

```diff
@@ -0,0 +1,734 @@
+                nodes.append(node)
+
+        return self.graph.get_subgraph(nodes)
+
+    def _get_involved_variables_from_node_list(
+        self, node_list
+    ) -> Dict[str, Set[str]]:
+        """
+        How each variables are involved within these nodes.
+
+        Return a dictionary with following keys
+        dependent_variables : union of all dependent_variables of each node
+        assigned_variables : union of all assigned_variables of each node
+        all_variables : union of dependent_variables and assigned_variables
+        input_variables : all_variables - assigned_variables
+        """
+
+        involved_variables = dict()
+        involved_variables["dependent_variables"] = set(
+            itertools.chain.from_iterable(
+                [
+                    self.node_context[nid]["dependent_variables"]
+                    for nid in node_list
+                ]
+            )
+        )
+        # variables got assigned within these nodes
+        involved_variables["assigned_variables"] = set(
+            itertools.chain.from_iterable(
+                [
+                    self.node_context[nid]["assigned_variables"]
+                    for nid in node_list
+                ]
+            )
+        )
+        # all variables within these nodes
+        involved_variables["all_variables"] = involved_variables[
+            "dependent_variables"
+        ].union(involved_variables["assigned_variables"])
+        # required input variables
+        involved_variables["input_variables"] = (
+            involved_variables["all_variables"]
+            - involved_variables["assigned_variables"]
+        )
+
+        # Add user defined parameter in to input variables list
+        for nid in node_list:
+            if self.input_parameters_node.get(nid, None) is not None:
+                involved_variables["input_variables"].add(
+                    self.input_parameters_node[nid]
+                )
+
+        return involved_variables
+
+    def _get_graph_segment_from_artifact_context(
+        self, artifact_context
+    ) -> GraphSegment:
+        """
+        Create the graphsegment from the artifact context
+
+        Note: might be able get rid of artifact_context entirely with some more
+        refactoring.
+        """
+        subgraph = self._get_subgraph_from_node_list(
+            artifact_context["node_list"]
+        )
+        involved_variables = self._get_involved_variables_from_node_list(
+            artifact_context["node_list"]
+        )
+        return GraphSegment(
+            artifact_name=artifact_context["artifact_name"],
+            graph_segment=subgraph,
+            segment_type=artifact_context["artifact_type"],
+            all_variables=involved_variables["all_variables"],
+            input_variables=involved_variables["input_variables"],
+            return_variables=artifact_context["return_variables"],
+        )
+
+    def _slice_session_artifacts(self) -> None:
```

Comment:
This function is large; would be easier to read if you split to different functions.